File: JAVA/34 - Find First and Last Position of Element in Sorted Array.java
/*
 34. Find First and Last Position of Element in Sorted Array (M)
 
 Approach 1 : BS
 BS the list twice, one find the left boundary, one find the right boundary.
 Time: O(2logn)
 Space: O(1)

*/

/* Approach 1 */
class Solution {
    public int[] searchRange(int[] nums, int target) {
        if (nums.length == 0) return new int[] {-1, -1};
        int[] index = new int[2];
        int l = 0, r = nums.length - 1, m = 0;
        
        // Find left boundary
        while (l < r - 1) {
            m = l + (r - l) / 2;
            if (nums[m] >= target) r = m;
            else l = m;
        }
        
        if (nums[l] != target && nums[r] != target) 
            return new int[] {-1, -1};
        
        if (nums[l] == target) index[0] = l;
        else index[0] = r;
        
        // Find right boundary
        l = 0;
        r = nums.length - 1;
        while (l < r - 1) {
            m = l + (r - l) / 2;
            if (nums[m] <= target) l = m;
            else r = m;
        }
        
        // If there are only two nodes in the list, then handle the r node first
        if (nums[r] == target) index[1] = r;
        else index[1] = l;
        
        return index;
    }
}

File: C++/34-Find_First_and_Last_Position_of_Element_in_Sorted_Array.cpp
/*
 34. Find First and Last Position of Element in Sorted Array
 Medium
*/

class Solution {
public:
    void func(vector<int>& nums, int l, int r, int target, int& ans_left, int& ans_right)
    {
        int index = -1, left = l, right = r;
        while (l <= r)
        {
            int m = l + (r - l) / 2;
            if (nums[m] == target)
            {
                index = m;
                break;
            }
 
            if (nums[m] < target)
                l = m + 1;
            else
                r = m - 1;
        }
 
        if (index == -1)
            return;
 
        if (index < ans_left)
            ans_left = index;
        if (index > ans_right)
            ans_right = index;
 
        func(nums, left, index - 1, target, ans_left, ans_right);
        func(nums, index + 1, right, target, ans_left, ans_right);
    }
 
    vector<int> searchRange(vector<int>& nums, int target) {
        int l = nums.size(), r = -1;
        func(nums, 0, nums.size() - 1, target, l, r);
        if (l > r)
            return {-1, -1};
        return {l, r};
    }
};
